lib/sql.js

```diff
@@ -156,200 +228,892 @@ SqlConnector.prototype.tableEscaped = function (model) {
+ */
+SqlConnector.prototype.fromRow = SqlConnector.prototype.fromDatabase = function(model, rowData) {
+  if (rowData == null) {
+    return rowData;
+  }
+  var props = this.getModelDefinition(model).properties;
+  var data = {};
+  for (var p in props) {
+    var columnName = this.column(model, p);
+    // Load properties from the row
+    var columnValue = this.fromColumnValue(props[p], rowData[columnName]);
+    if (columnValue !== undefined) {
+      data[p] = columnValue;
+    }
+  }
+  return data;
+};
+
+/**
+ * Find matching model instances by the filter
  *
  * @param {String} model The model name
- * @param {Function} callback The callback function
- * @param {Object} where The where clause
+ * @param {Object} filter The filter
+ * @param {Function} [cb] The cb function
  */
-SqlConnector.prototype.count = function count(model, callback, where) {
+SqlConnector.prototype.all = function find(model, filter, options, cb) {
   var self = this;
-
-  var whereClause = buildWhere(self, model, where);
-  this.queryOne('SELECT count(*) as cnt FROM ' +
-    this.tableEscaped(model) + ' ' + whereClause, function (err, res) {
+  // Order by id if no order is specified
+  filter = filter || {};
+  var stmt = this.buildSelect(model, filter, options);
+  this.execute(stmt.sql, stmt.params, options, function(err, data) {
     if (err) {
-      return callback(err);
+      return cb(err, []);
+    }
+
+    var objs = data.map(function(obj) {
+      return self.fromRow(model, obj);
+    });
+    if (filter && filter.include) {
+      self.getModelDefinition(model).model.include(objs, filter.include, cb);
+    } else {
+      cb(null, objs);
     }
-    callback(err, res && res.cnt);
   });
 };
+SqlConnector.prototype.findAll = SqlConnector.prototype.all;
 
 /**
- * Update attributes for a given model instance
- * @param {String} model The model name
+ * Find by id
+ * @param {String} model The Model name
  * @param {*} id The id value
- * @param {Object} data The model data instance containing all properties to be updated
+ * @param {Object} options The options object
  * @param {Function} cb The callback function
  */
-SqlConnector.prototype.updateAttributes = function updateAttributes(model, id, data, cb) {
-  if (!isIdValuePresent(id, cb)) {
+SqlConnector.prototype.findById = SqlConnector.prototype.find = function(model, id, options, cb) {
+  if (id == null) {
+    process.nextTick(function() {
+      var err = new Error('id value is required');
+      err.statusCode = 400;
+      if (cb) {
+        cb(err);
+      }
+    });
     return;
   }
-  var idName = this.getDataSource(model).idName(model);
-  delete data[idName];
+  var self = this;
   var where = {};
+  var idName = this.idName(model);
   where[idName] = id;
-  this.updateAll(model, where, data, cb);
+
+  var filter = {limit: 1, offset: 0, order: idName, where: where};
+  return this.all(model, filter, options, function(err, results) {
+    cb(err, (results && results[0]) || null);
```

Comment:
BTW it would be great if we started to use one of the following two function prologues instead of checking `cb` value everywhere we call it.

```
function(cb) {
  // (1) cb is required
  assert(cb, 'callback is required');
  // (2) cb is optional
  cb = cb || function(){};

  // method body, we can rely on cb being always defined
}
```

Here are the benefits of this is solution as I see them:
 - No space for unintended bugs caused by forgotten `cb &&` check
 - Function arguments are checked in one place, at the function start. Instead of sprinkling argument checks all over the function body.